### DISCUSSION TITLE
Conventions for macros

### DISCUSSION THREAD

**Discussion by v--** [Clopen-Analysis/latex#11] [General]
This is a suggestion for having conventions for macros. The problem that I see with the current macros is that their names are not predictable - for example, `\NN` is blackboard bold N and `\BB` is calligraphic bold.

Here are some suggestions based on my personal experience:

* I prefix blackboard bold letters with `\B` so that the natural numbers become `\BN`, the complex numbers become `\BC` and small k becomes `\Bk`. Analogously, I prefix calligraphic letters with `\C`.

* I like to use boldface for categories so rather than introducing different macros for every category I set up a macro `\Cat` and then write `\Cat{Set}` or `\Cat{Vect}_{\Ck}`. I did similar things with probability distributions in another document where I used `\D{Norm}` and `\D{Cauchy}`.

* I like to use mathtools delimiters that allow you to define `\DeclarePairedDelimiter \Paren \lparen \rparen` and later use `\Paren*{something}` rather than `\left( something \right)` or `\Paren[\Big]{something}` instead of `\Big( something \Big)`.

Also, I think the project would benefit from having separate "sty" files, especially if at some point it is to produce more than one PDF file.

**Comment by qwinters** [comment]
Coming at this from the perspective of wanting to push TeX document content onto individual webpages on the website I would be in-favor of aggressively minimizing or eliminating macros completely. Macros are just not fun to maintain across MathJax and various TeX documents. 

Stylistically, I think macros make it significantly harder for people to contribute to a communal TeX code base since everyone has different preferences for macros. If I'm reading TeX code and have to constantly flip back to the macros section to read what the macros are, I'm a lot less inclined to spend the time to try to write up a PR to add content. 

To take your example, from my experience, the traditionally `mathbb` symbols (R, C, Z, N, Q, E, P, S, O, D, H, T, 1) pretty much never overlap with the traditionally `mathcal` letters (B, L, F, G, A, M, U, W, V) in the analysis/probability version of "traditionally" (with the exception of maybe A with `\mathbb{A}` used for affine spaces in convexity theory and `\mathcal{A}` sometimes used for a collection or index for a net for example).  If I were to see `\BR` in the code, I would assume that your code is refering to something else (e.g. it might a macro for denoting the standard brownian motion as `B_t` or denoting the unit ball like `\mathbb{B}_{\mathbb{R}}`. 

Maybe there's a good compromise to be made with macros to shorten some existing functions (e.g. `\bb{R}` instead of `\mathbb{R}`), but I think it should be minimal and so obvious that you do not have to look back up at the macros portion of the file to figure it out. 

> mathtools delimiters 

For the delimiter matching, everything you could need is already defined in the `physics` or `braket` packages (which are natively supported by MathJax) E.g. use `\pqty{...}` for enclosing a quantity in parenthesis, `\Set{...}` for defining sets, etc.

We should 100% prefer these to `mathtools` to maintain compatibility with MathJax for publishing web-versions of the document contents, regardless on the ultimate decision on macros.

**Reply by abackus** [reply]
Note that Stacks Project seems to have no problem with macros, c.f. https://github.com/stacks/stacks-project/blob/master/preamble.tex. I do agree that we shouldn't just make a ton of macros for the sake of making them, as that just makes the code hard to read, but it's not clear to me that just because this should be publishable on a website means that we need to forsake macros.

**Reply by v--** [reply]
You have good points, especially regarding the braket package, and I'm all for less custom macros myself. My suggestions were based on the existing code. I saw that \left( was used in a few places so I suggested mathtools (although your suggestion is indeed better for the use case, delimiters are still useful for things like norms). I saw that there are a few category macros so I suggested a single macro for all of them. I saw that there are some letter shortcuts so I suggested a way for them to grow in quantity without eventually overlapping and without much cognitive overhead - it is enough to remember a two-letter macro starting eith C is calligraphic. Based on the current convention, \AA may either be blackboard bold or calligraphic and it is not possible to know in general without looking at the definition (although it may be clear from the context).

Anyway, these are just suggestions after all. I don't really mind any particular style. It is indeed perhaps best to see how the project grows prior to making stylistic decisions about macros.

**Reply by qwinters** [reply]
> Note that Stacks Project seems to have no problem with macros

While this is true, there seems to be a lot of work to get it done in the [web book](https://github.com/stacks/stacks-project/blob/master/scripts/web_book.py) script and all that work seems ad-hoc and I personally at least don't want to maintain or worry about the script breaking when the alternative is just native compatibility with MathJax.

But if there's a cleaner way to keep macros, then I'm not _a priori_ against them but still in favor of minimizing their use for stylistic reasons. 

> delimiters are still useful for things like norms 

For norms and absolute values the physics package already has `\norm{}` and `\abs{}` built in and they work pretty much how you would want (e..g `\norm{X}_{\mathcal{L}^p}` would gives you the proper L^p subscript for the 1-Wasserstein norm). If you haven't used the physics package before I would highly recommend taking a look at [the documentation](http://mirrors.ibiblio.org/CTAN/macros/latex/contrib/physics/physics.pdf) -- aside from abbreviations for faster typing, it has pretty much every complicated macro/function you could need for analysis.

> without eventually overlapping and without much cognitive overhead - it is enough to remember a two-letter macro starting eith C is calligraphic

I'm just saying to some people (like myself) the prefixed macros convention proposed is equally as confusing as the double letter `\RR`type convention. 

I personally have moved away from using either in my personal LaTeX work (though partially to maintain easy copy/paste) and have opted to go for vim snippets / tex-studio tab autocomplete for speeding up typing. 

It seems like from a quick search through the [Stacks repo](https://github.com/stacks/stacks-project/search?q=mathbf) that they have also opted for using `\mathbb{}`, `\mathbf{}`, and `\mathcal{}` directly in the text (which I think is a good reason to adopt, since they've been through this before and opted for that style, so why wouldn't it work for this project?)

**Reply by v--** [reply]
I knew about the physics package but I haven't supposed that it is so helpful. Will definitely try it out now.

Also, this project has about 500 references to mathbb letters in about 150 pages, which seems to be much more than all math font modifiers in the Stacks project. I think that at some point using \mathbb, \mathcal and so forth just becomes too cumbersome to write manually. As far as I know, MathJax has no problems with some simple plain TeX macros like these ones.

**Reply by v--** [reply]
I actually tried `physics` and, as nice as it seems, it has more shortcomings than the average package. For example, its flexible (and very special) syntax is sometimes too flexible and produces weird errors if you do not use the correct brackets. It's maybe not the most major problem if you are used to it but it's not exactly a nice situation to be in. The problem is a little deeper than that because `physics` overloads some of the builtin macros.

For example, the following does not compile:
```
\documentclass{article}

\usepackage{physics}

\begin{document}
  \begin{equation*}
    \norm x
  \end{equation*}
\end{document}
```
and changing `\norm x` to `\norm{x}` compiles just fine.

[This thread](https://tex.stackexchange.com/questions/471563/physics-replacement-effort) has some discussion on what alternatives there are to the physics package. I'm not sure what is best for this project, but I would still prefer mathtools delimiters for my personal projects.

**Reply by qwinters** [reply]
I have never had an issue with physics package in my testing, as long as you code with unambiguous style (which you should be because otherwise it's hard to read your code). 

E.g. in the example you mentioned with `\norm x` vs `\norm{x}`, I would immediately reject a PR that uses the former instead of latter. How is someone reading your TeX supposed to differentiate between 
```LaTeX
\norm{x} +\frac{\delta}{2}
```
and 
```LaTeX
\norm{x +\frac{\delta}{2}}
```
unless you use the `{...}` to give norm its arguments. This is the same convention that I would expect from any programming language. Automatic argument inference is just not a thing any compiler will do well due to how grammars of programming languages are built. 

Do you have other examples of things that break that would be bigger issues?

I took a look through the thread (and the ones linked in the comments, and the ones linked in the comments on those threads) and the only thing I could see cited as a particular thing that seemed hard to deal with was a small amount of whitespace where someone didn't want it, which I don't think is something we should care much about. 

The functions that physics overrides are changed only in the most trivial of ways (e.g. you can use `\sin(x + \frac{\pi}{4})` instead of `\sin\left(x + \frac{\pi}{4}\right)`. Using `\sin x` or `\sin^2 x` for these functions still works as expected). Also physics still provides access to those functions via their full names (e.g. `\cosine` gives the AMS version).

**Reply by v--** [reply]
I understand your point regarding braces but the example is a bit ridiculous - it's clear from the context that `x` is a vector and `\delta` is a scalar. Rejecting a pull request because somebody has a different style is immature, to say the least. There's a reason we have and use shorthands in programming languages and TeX in particular. It's also conventional for macros to have a fixed amount of mandatory arguments so you know the exact result of `\frac 1 2` and don't have to write `\frac{1}{2}` every time.

Since you started speaking about programming languages, from my programming experience I've learned not to rely on obscure libraries with shady reputation, especially ones that attempts to do too much metaprogramming. I'll use the `physics` package if I contribute to existing code that uses it but I'm also not going to introduce it myself.

This discussion begins to delve into some gourmet bikeshedding so I suggest that whoever does the pretext/mathjax/whatever work should be the one who makes any decisions regarding typesetting balanced braces. I'm in favor of avoiding manually writing `\left` and `\right` everywhere but would use whatever solution turns out to work the best.

**Comment by v--** [comment]
By the way, I just found out that the `unicode-math` package has predefined macros for script (`\mscrX` - note that in unicode and fonts like STIX, script and calligraphic letters are the same), fraktur (`\mfracX`) and blackboard (`\BbbX`) fonts (and a few others). [This MathJax plugin](https://github.com/AmerMathSoc/mathjax-unicode-math) seems to support these macros. The advantage over e.g. `\mathbb` is that the produced PDF uses actual unicode symbols rather than simply another font. I assume the same holds for web pages.

This can also be achieved by using `\symbb` rather than `\mathbb` and analogously for `\symfrak` and `\symsrc`/`symcal`.

In other words, both `\symbb{R}` and `\BbbR` should be more unicode-compatible (and hence copy-paste-compatible) than using plain `\mathbb{R}` or any of its macros. Furthermore, this alleviates any need for macros like `\RR`.
